src/ray/core_worker/store_provider/memory_store/memory_store.cc

```diff
@@ -0,0 +1,206 @@
+#include <condition_variable>
+#include "ray/common/ray_config.h"
+#include "ray/core_worker/context.h"
+#include "ray/core_worker/core_worker.h"
+#include "ray/core_worker/object_interface.h"
+#include "ray/core_worker/store_provider/memory_store_provider.h"
+
+namespace ray {
+
+/// A class that represents a `Get` reuquest.
+class GetRequest {
+ public:
+  GetRequest(const std::vector<ObjectID> &object_ids, bool remove_after_get);
+
+  const std::vector<ObjectID> &ObjectIds() const;
+
+  /// Wait until all requested objects are available, or timeout happens.
+  ///
+  /// \param timeout_ms The maximum time in milliseconds to wait for.
+  /// \return Whether all requested objects are available.
+  bool Wait(int64_t timeout_ms);
+  /// Set the object content for the specific object id.
+  void Set(const ObjectID &object_id, std::shared_ptr<RayObject> buffer);
+  /// Get the object content for the specific object id.
+  std::shared_ptr<RayObject> Get(const ObjectID &object_id) const;
+  /// Whether this is a `get` request.
+  bool ShouldRemoveObjects() const;
+
+ private:
+  /// Wait until all requested objects are available.
+  void Wait();
+
+  /// The object IDs involved in this request. This is used in the reply.
+  const std::vector<ObjectID> object_ids_;
+  /// The object information for the objects in this request.
+  std::unordered_map<ObjectID, std::shared_ptr<RayObject>> objects_;
+
+  // Whether the requested objects should be removed from store
+  // after `get` returns.
+  const bool should_remove_;
+  // Whether all the requested objects are available.
+  bool is_ready_;
+  mutable std::mutex mutex_;
+  std::condition_variable cv_;
+};
+
+GetRequest::GetRequest(const std::vector<ObjectID> &object_ids, bool remove_after_get)
+    : object_ids_(object_ids), should_remove_(remove_after_get) {}
+
+const std::vector<ObjectID> &GetRequest::ObjectIds() const { return object_ids_; }
+
+bool GetRequest::ShouldRemoveObjects() const { return should_remove_; }
+
+bool GetRequest::Wait(int64_t timeout_ms) {
+  if (timeout_ms < 0) {
```

Comment:
Is this documented somewhere? It seems surprising to me!

Reply:
I'm using < 0 to indicate wait indefinitely, like -1. 

Currently raylet wait has this check:
`RAY_CHECK(timeout_ms >= 0 || timeout_ms == -1);`

I think it's also fine to use that behavior here. What's your suggestion then?

Reply:
will change to the ray check above and only treat -1 as indefinite. Let me know if this is ok to you.

Reply:
only treating -1 as indefinite would be clearer I think yeah

Reply:
ok!